src/librustc_data_structures/interner.rs

```diff
@@ -0,0 +1,291 @@
+// Copyright 2018 The Rust Project Developers. See the COPYRIGHT
+// file at the top-level directory of this distribution and at
+// http://rust-lang.org/COPYRIGHT.
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+use std::marker::PhantomData;
+use std::hash::Hash;
+use std::hash::Hasher;
+use std::hash::BuildHasher;
+use std::mem::{self, size_of};
+use std::ptr::{Unique, NonNull};
+use std::alloc::{Global, Alloc};
+use std::collections::hash_map::RandomState;
+use std::borrow::Borrow;
+use std::fmt;
+
+const ENTRIES_PER_GROUP: usize = 5;
+
+#[repr(align(64), C)]
+pub struct Group {
+    hashes: [u32; ENTRIES_PER_GROUP],
+    size: u32,
+    values: [u64; ENTRIES_PER_GROUP],
```

Comment:
Have you tried this on a 32-bit system? Is there a reason not to make this generic? As in:
```rust
struct Group<T: Sized> {
    hashes: [u32; ENTRIES_PER_GROUP],
    size: u32,
    values: [*const T; ENTRIES_PER_GROUP],
}
```